### PR TITLE
Implement PEP 498 string interpolation

### PR DESCRIPTION
**Pull request by kirbyfan64**
This implements [Python PEP 498 string interpolation](https://www.python.org/dev/peps/pep-0498/). [Again](https://github.com/howl-editor/howl/pull/117). The syntax is more complicated than it looks!

### CODE REVIEW

**Comment by nilnor** [comment]
@shalabhc is this something you have or could look at, given that you're more of a Python buff than yours truly?

**Comment by shalabhc** [comment]
It would be great to have these format strings styled. I've looked at it briefly and will review more thoroughly. For now here are some issues:

f'{value:#format}' should not style '#format' as a comment.
f'{value:def}' should not style 'def' as keyword.
f'{value:>+#10.2)' is not styled correctly.

This seems to handle the built in format specification well. Handing everything properly is quite a challenge. Also, given that the language allows 'Each formattable type may define how the format specification is to be interpreted' (https://docs.python.org/2/library/string.html#formatspec) - I wonder if it's useful to handle the numeric format specifications at all? Note that the format spec can be absolutely any string (including my examples above). 

I'm thinking that anything after the ':' - i.e. the entire format spec - could be styled as a single style, maybe 'special'. While this would lose the special style for numeric format strings, it will considerably simplify the implementation and be more correct in the general sense. However, this will also lose the style for nested {expression} in the format spec. Maybe that is OK - what do you think @kirbyfan64?

**Comment by kirbyfan64** [comment]
@shalabhc Hmm...you're probably right...I think I may be able to save the formatting of brackets, too. Hold on a sec...

**Comment by kirbyfan64** [comment]
@shalabhc I pushed a new commit...but is it possible with LPeg to match any character? I tried replacing [this line](https://github.com/howl-editor/howl/commit/784f4455cae07d1167d658779608012479d79a9e#diff-851d9b84896270cc7e3bbea3014007a5R97) with:

```moonscript
format_spec = c('operator', ':') * (format_number + c 'special', P 1)^1
```

[like the LPeg docs for `P` say](http://www.inf.puc-rio.br/~roberto/lpeg/#op-p), but it didn't match anything.

**Comment by shalabhc** [comment]
@kirbyfan64 - not sure. Sorry my knowledge of LPeg isn't great but I'll poke around a bit and see if I can figure this out.

**Comment by shalabhc** [comment]
@kirbyfan64 - So I think instead of matching any char, it might be better to match one or more chars that are *not* `{` - what do you think? It might be easier to implement too.

**Comment by kirbyfan64** [comment]
@shalabhc All issues addressed! I also made the formatting a bit more lenient to avoid string formatting getting screwed in the middle of typing.

**Comment by kirbyfan64** [comment]
Bump?

**Comment by shalabhc** [comment]
Sorry for the delay on this. I tried this out but in `f'{value:#format}'` I still get '#format' styled as a comment.

Some other implementation comments inline.

File: bundles/python/python_lexer.moon [hunk 1]
```diff
@@ -100,4 +132,24 @@ howl.aux.lpeg_lexer ->
       identifier,
       decorator
     }
+
+    tsq_interpolation: gen_string_interpolation 'ts'
```
Comment (shalabhc) [inline bundles/python/python_lexer.moon]: It would be nice to not have all variants of the `f` strings here. Instead just call these, say,  `f_string` as a single rule that is defined separately above using `any`. Also, style wise, it seems strange to use interpolation with 'ts' etc. to refer to other rules - it makes it somewhat hard to follow.  Would be possible to just generate all the variants of the f strings by iterating over the names of the normal strings? Maybe you could add the generated rules to a list and then make an `any` rule out of it?

**Comment by kirbyfan64** [comment]
@shalabhc Are you sure? I just tested it, and it's styled correctly.

File: bundles/python/python_lexer.moon (hunk 1 above)
Comment (kirbyfan64) [inline bundles/python/python_lexer.moon]: @shalabhc I just updated it. Is it better now?

File: bundles/python/python_lexer.moon [hunk 2]
```diff
@@ -100,4 +140,14 @@ howl.aux.lpeg_lexer ->
...
+
+    f_string: any f_strings
+
+    string: any {
+      f_prefix * V'f_string'
+      string
+    }
   }
+
+  rules[k] = v for k, v in pairs interpolations
```
Comment (shalabhc) [inline bundles/python/python_lexer.moon]: I'm not sure why it is necessary to inject all interpolations into the main rule set? Since `f_string` is already in `all`, wouldn't that just work?

Comment (kirbyfan64) [inline bundles/python/python_lexer.moon]: @shalabhc Because the rules recurse into each other with `V`. Just like in the MoonScript one.

**Comment by shalabhc** [comment]
@kirbyfan64 - ah you're right `f'value:#format'` is fixed. However I found the following issues:

```python

f'{value:abc{f()}}' # not good, string not closed?

f'{val!#comment}' # not good, styled as comment, should be error or unstyled ?

f'{a + b}' # not good, string not closed?

```

**Comment by kirbyfan64** [comment]
@shalabhc *sigh* I'm pretty sure I fixed all the issues this time. If all's OK now, I can squash the commits.

**Comment by shalabhc** [comment]
@kirbyfan64 - seems to work pretty well but I found a special case ;)

```
f'{x!=1}'
```

I think the `!=` operator has special exception in the pep. If fixing this is too complex, I'm ok with leaving it as is.

**Comment by kirbyfan64** [comment]
Crud, my rebase killed the commit history. Hold on...
